nemo/collections/nlp/data/text_classification/text_classification_dataset.py

```diff
@@ -60,193 +60,215 @@ def output_types(self) -> Optional[Dict[str, NeuralType]]:
+        """collate batch of input_ids, segment_ids, input_mask, and label
+        Args:
+            batch:  A list of tuples of (input_ids, segment_ids, input_mask, label).
         """
-
-        self.features = []
-        for sent_id in range(len(all_sent_subtokens)):
-            sent_subtokens = all_sent_subtokens[sent_id]
-            label = labels[sent_id]
+        max_length = 0
+        for input_ids, segment_ids, input_mask, label in batch:
+            if len(input_ids) > max_length:
+                max_length = len(input_ids)
+
+        padded_input_ids = []
+        padded_segment_ids = []
+        padded_input_mask = []
+        labels = []
+        for input_ids, segment_ids, input_mask, label in batch:
+            if len(input_ids) < max_length:
+                pad_width = max_length - len(input_ids)
+                padded_input_ids.append(np.pad(input_ids, pad_width=[0, pad_width], constant_values=self.pad_id))
+                padded_segment_ids.append(np.pad(segment_ids, pad_width=[0, pad_width], constant_values=self.pad_id))
+                padded_input_mask.append(np.pad(input_mask, pad_width=[0, pad_width], constant_values=self.pad_id))
+            else:
+                padded_input_ids.append(input_ids)
+                padded_segment_ids.append(segment_ids)
+                padded_input_mask.append(input_mask)
+            labels.append(label)
+
+        return (
+            torch.LongTensor(padded_input_ids),
+            torch.LongTensor(padded_segment_ids),
+            torch.LongTensor(padded_input_mask),
+            torch.LongTensor(labels),
+        )
+
+    @staticmethod
+    def get_features(all_sents, tokenizer, max_seq_length, labels=None, verbose=True):
+        """Encode a list of sentences into a list of tuples of (input_ids, segment_ids, input_mask, label)."""
+        features = []
+        sent_lengths = []
+        too_long_count = 0
+        for sent_id, sent in enumerate(all_sents):
+            sent_subtokens = [tokenizer.cls_token]
+            for word in sent:
+                word_tokens = tokenizer.text_to_tokens(word)
+                sent_subtokens.extend(word_tokens)
+
+            if max_seq_length > 0 and len(sent_subtokens) + 1 > max_seq_length:
+                sent_subtokens = sent_subtokens[:max_seq_length]
+                too_long_count += 1
+
+            sent_subtokens.append(tokenizer.sep_token)
+            sent_lengths.append(len(sent_subtokens))
 
             input_ids = [tokenizer.tokens_to_ids(t) for t in sent_subtokens]
 
             # The mask has 1 for real tokens and 0 for padding tokens.
             # Only real tokens are attended to.
             input_mask = [1] * len(input_ids)
+            segment_ids = [0] * len(input_ids)
 
-            # Zero-pad up to the sequence length.
-            while len(input_ids) < max_seq_length:
-                input_ids.append(0)
-                input_mask.append(0)
-            segment_ids = [0] * max_seq_length
```

Comment:
there is no padding now at all...?

Reply:
Not here when we process the samples. Previously all samples were padded to max_seq_length. Now we have a collate_fn which does the padding to the max length in a batch.

Reply:
ooh, missed that one. Well, that once again brings up the question: what should:
1) the preprocessing in "external script" vs
2) vs preprocessing in PTL model.data_prepare() vs
3) preprocessing when dataset instance is created vs
4) preprocessing in collate_fn?
do?

Reply:
Currently 
+we convert other formats to NeMo format with pre-processing and conversion scripts.
+We do not have data_prepare()
+dataset would tokenize and make ready the input to the model, the exact tensors to be passed to the forward. Then store all of them in a list ready or store them in a cache file.
+collate_fn would just concat some samples with padding and pass them